Fix upgrader for some cases

cpunion opened the pull request:
See https://github.com/rmosolgo/graphql-ruby/issues/1177

- [x] NameError (uninitialized constant BaseObject)
- [x] Generate classes base on Types::Base* instead of Base*
- [x] Convert `input_field` to `argument`, convert `null:` to `required:`
- [x] Convert types base on InputObjectType, ScalarType
- [x] Convert `interfaces [...]` to `implemnts`
~- [ ] Convert `description`~
~- [ ] Allow definition of `resolve` within field block~

Willianvdv commented on lib/graphql/upgrader/member.rb:
```diff
@@ -39,16 +40,25 @@ def upgrade
-              if may_return_null
-                f += ', null: true'
+              if is_argument
+                if may_return_null
```
Now that `may_return_null` also determines if an input is mandatory, maybe `has_bang` is a more appropriate name.

cpunion replied:
Updated

Willianvdv commented on lib/graphql/upgrader/member.rb:
```diff
@@ -127,7 +135,23 @@ def rename_property_to_method(transformable)
       attr_reader :member
+
+      private
+        GRAPHQL_TYPES = '(Object|InputObject|Interface|Enum|Scalar|Union)'
```
👏

Willianvdv commented:
Great work @cpunion!

Willianvdv commented:
>  Convert description

Have you seen #1175?

cpunion commented:
@Willianvdv #1175 is good. So type, description and resolve can be the same way. I tried parse ast to convert them, but not clear, for this "disposable tool".

Willianvdv commented:
> I tried parse ast to convert them, but too complex for this one time tool.

Initially I also went for the AST transformation route, and like you, I found it too hard to get it right! 😆

Willianvdv commented on lib/graphql/schema/field.rb:
```diff
@@ -132,6 +132,11 @@ def description(text)
           end
         end
+
+        def resolve(resolve_proc)
```
Since this is a change in the 1.8 API, we should have a separate PR to discuss it (like @esjee did for #1158). Splitting this into two separate PRs allows us to focus on the upgrader in this PR and have a focused discussion around the API change. Please note: I don't object to this change per se, but I do want to stick to the API change process.

What do you think?

Willianvdv replied:
Besides this comment, I really like the improvements on the upgrader!

cpunion replied:
OK. I will pick it to a new PR.

cpunion commented:
Updated, removed last commit about resolve in field block.

cpunion commented:
I have done an ast transformer, the following is output.

Test result:
<details>
20:24:35 - INFO - Running: spec/graphql/upgrader/member_spec.rb
warning: parser/current is loading parser/ruby24, which recognizes
warning: 2.4.0-compliant syntax, but you are running 2.4.1.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.

# Running tests with run options --guard --seed 50081:

FF.F..F.....

Finished tests in 0.070509s, 170.1910 tests/s, 411.2950 assertions/s.


Failure:
GraphQL::Upgrader::Member::implements#test_0001_upgrades interfaces to implements [/Users/lijie/source/graphql-ruby/spec/graphql/upgrader/member_spec.rb:10]
Minitest::Assertion: --- expected
+++ actual
@@ -1,9 +1,13 @@
 "
 " +
+"
+" +
 "        implements Types::SearchableType
 " +
 "        implements Types::CommentableType
 " +
+"
+" +
 "        implements Types::ShareableType
 " +
 "      "


Failure:
GraphQL::Upgrader::Member::name#test_0001_removes the name field if it can be inferred from the class [/Users/lijie/source/graphql-ruby/spec/graphql/upgrader/member_spec.rb:10]
Minitest::Assertion: --- expected
+++ actual
@@ -2,6 +2,8 @@
 " +
 "        class UserType < Types::BaseObject
 " +
+"
+" +
 "        end
 " +
 "      "


Failure:
GraphQL::Upgrader::Member::name#test_0002_removes the name field if it can be inferred from the class and under a module [/Users/lijie/source/graphql-ruby/spec/graphql/upgrader/member_spec.rb:10]
Minitest::Assertion: --- expected
+++ actual
@@ -2,6 +2,8 @@
 " +
 "        class Types::UserType < Types::BaseObject
 " +
+"
+" +
 "        end
 " +
 "      "


Failure:
GraphQL::Upgrader::Member::fields#test_0001_upgrades to the new definition [/Users/lijie/source/graphql-ruby/spec/graphql/upgrader/member_spec.rb:10]
Minitest::Assertion: --- expected
+++ actual
@@ -2,6 +2,8 @@
 " +
 "        field :name, -> { String }, null: true do
 " +
+"
+" +
 "        end
 " +
 "      "


12 tests, 29 assertions, 4 failures, 0 errors, 0 skips

20:24:40 - INFO - Inspecting Ruby code style: lib/graphql/upgrader/member.rb
Inspecting 1 file
.
</details>


It can converts `type`, `description`, `resolve` within field block, but generates some new lines and spaces. So if test it ignore empty lines and spaces, all tests past:

```ruby
  def assert_equal_ignore_spaces(a, b)
    a = a.split("\n").map{|line| line.strip}.select{|line| !line.empty?}.join("\n")
    b = b.split("\n").map{|line| line.strip}.select{|line| !line.empty?}.join("\n")
    assert_equal(a, b)
  end
```

I think it can do more than RegExp version. But I have no enough time to refactor. If you are interested, I can post a draft PR.

BTW: I found some errors in test cases when test this upgrader, should fix them:

https://github.com/rmosolgo/graphql-ruby/blob/1.8-dev/spec/graphql/upgrader/member_spec.rb#L230

https://github.com/rmosolgo/graphql-ruby/blob/1.8-dev/spec/graphql/upgrader/member_spec.rb#L245

Willianvdv commented:
Could you post the PR or post a gist, I'm curious how you implemented it!

rmosolgo commented:
Thanks for fixing these cases, and for opening the AST-based PR. I'm interested to review it soon!
